src: kw_remote: Add support for global remote.config

Kamorst opened the pull request:
`kw remote` checks for the presence of `$PWD/.config/kw/remote.conf` and use it if a local one is not found.

Fixes #775.

Kamorst commented:
TODO: add some more tests.

davidbtadokoro commented:
> TODO: add some more tests.

Implemented single tests for add/remove/rename/list/set-default global remote.

rodrigosiqueira commented on src/kw_remote.sh:
```diff
@@ -43,24 +44,59 @@ function remote_main()
 {
-  local remote_config
-  local str_process
+  local has_local_config="false"
+  local has_global_config="false"
```
Uses single-quotes.

rodrigosiqueira commented on src/kw_remote.sh:
```diff
@@ -43,24 +44,59 @@ function remote_main()
-    complain 'You do not have .kw folder, try kw init first'
-    return 22 # EINVAL
+  if [[ -d "${PWD}/.kw" ]]; then
+    has_local_config="true"
```
Use single quotes around true.

rodrigosiqueira commented on src/kw_remote.sh:
```diff
@@ -43,24 +44,59 @@ function remote_main()
+    has_local_config="true"
+  fi
+
+  if [[ "${has_local_config}" = "false" && -d "${KW_ETC_DIR}" ]]; then
```
Use single quotes and `-eq` instead of `=`.

rodrigosiqueira commented on src/kw_remote.sh:
```diff
@@ -43,24 +44,59 @@ function remote_main()
-  if [[ ! -f "${local_remote_config_file}" ]]; then
-    complain 'Did you run kw init? It looks like that you do not have remote.config'
+  if [[ "${has_local_config}" = "true" ]]; then
+    #say "Using local remote.config." 1>&2
+    printf '%s' "${local_remote_config_file}"
```
`say` or `print`?

rodrigosiqueira commented on src/kw_remote.sh:
```diff
@@ -43,24 +44,59 @@ function remote_main()
+  if [[ "${has_local_config}" = "true" ]]; then
+    #say "Using local remote.config." 1>&2
+    printf '%s' "${local_remote_config_file}"
+    return 0
+  elif [[ "${has_global_config}" = "true" ]]; then
+    #say "Using global remote.config." 1>&2
+    printf '%s' "${global_remote_config_file}"
+    return 0
```
Maybe move these two `return 0` after the `if` condition.

rodrigosiqueira commented on src/kw_remote.sh:
```diff
@@ -43,24 +44,59 @@ function remote_main()
+    printf '%s' "${global_remote_config_file}"
+    return 0
+  else
+    complain 'No local or global config found. Please reinstall kw or run "kw init" to create a .kw in this dir.' 1>&2
```
Why `1>&2` at the end of the complain?

rodrigosiqueira commented on src/kw_remote.sh:
```diff
@@ -43,24 +44,59 @@ function remote_main()
+
+  remote_config_file="$(choose_correct_remote_config_file)"
+  ret="$?"
+  [[ "$ret" != "0" ]] && return "$ret"
```
You don't need quotes around 0.

rodrigosiqueira commented on src/kw_remote.sh:
```diff
@@ -43,24 +44,59 @@ function remote_main()
+
+  if [[ ! -f "$remote_config_file" ]]; then
+    # If we don't have a remote.config file yet, let the user know
+    if [[ "$remote_config_file" = "$local_remote_config_file" ]]; then
```
replace `=` with `-eq`.

Kamorst replied:
i'm getting some strange error when making this alteration.
`$> ./run_tests.sh test kw_remote >/dev/null`
`./src/kw_remote.sh: line 87: [[: /tmp/shunit.IQPPtY/tmp/.kw/remote.config: syntax error: operand expected (error token is "/tmp/shunit.IQPPtY/tmp/.kw/remote.config")`
`./src/kw_remote.sh: line 97: /tmp/shunit.IQPPtY/tmp/.kw/remote.config: No such file or directory`
`./src/kw_remote.sh: line 87: [[: /tmp/shunit.IQPPtY/tmp/.config/kw/remote.config: syntax error: operand expected (error token is "/tmp/shunit.IQPPtY/tmp/.config/kw/remote.config")`

rodrigosiqueira replied:
I reread the bash documentation, and I realized that -eq is only for arithmetic cooperation. We can use `=`, but tbh, I think `==` is more friendly since other languages use it; could you replace `=` with `==`?

rodrigosiqueira commented on src/kw_remote.sh:
```diff
@@ -89,6 +125,11 @@ function add_new_remote()
+
+  remote_config_file="$(choose_correct_remote_config_file)"
+  ret="$?"
+  [[ "$ret" != "0" ]] && return "$ret"
```
You don't need quotes around 0.

rodrigosiqueira commented on src/kw_remote.sh:
```diff
@@ -43,24 +44,57 @@ function remote_main()
 }
 
+# Returns the local or global remote.config location.
+# The local has precedence.
```
Document this function return.

rodrigosiqueira commented on src/kw_remote.sh:
```diff
@@ -127,11 +169,11 @@ function add_new_remote()
 
   # Check if remote name already exists
-  grep -xq "^Host ${name}$" "$local_remote_config_file"
+  grep -xq "^Host ${name}$" "$remote_config_file"
```
Since you are changing this file, can you add one extra commit where you replace:

`grep -xq ..` with `grep --line-regexp --quiet`.

Using the full parameter name in the bash code makes it easier to understand what's going on in the line.

P.s.: I'm not sure if we documented this idea, but if not we should.

rodrigosiqueira commented on src/kw_remote.sh:
```diff
@@ -127,11 +169,11 @@ function add_new_remote()
-    sed -i -r "/^Host ${name}$/{n;s/Hostname.*/Hostname ${remote_parameters['REMOTE_IP']}/}" "$local_remote_config_file"
-    sed -i -r "/^Host ${name}$/{n;n;s/Port.*/Port ${remote_parameters['REMOTE_PORT']}/}" "$local_remote_config_file"
-    sed -i -r "/^Host ${name}$/{n;n;n;s/User.*/User ${remote_parameters['REMOTE_USER']}/}" "$local_remote_config_file"
+    sed -i -r "/^Host ${name}$/{n;s/Hostname.*/Hostname ${remote_parameters['REMOTE_IP']}/}" "$remote_config_file"
+    sed -i -r "/^Host ${name}$/{n;n;s/Port.*/Port ${remote_parameters['REMOTE_PORT']}/}" "$remote_config_file"
+    sed -i -r "/^Host ${name}$/{n;n;n;s/User.*/User ${remote_parameters['REMOTE_USER']}/}" "$remote_config_file"
```
same as my previous comment.

rodrigosiqueira reviewed (changes requested):
That is a very nice update!

I added a few comments inline.

Btw, we need one option to indicate kw remote if it should update the global config or the local one where local is the default if nothing is used.

For example:
```
kw remote add origin root@something --global # update global config
kw remote add origin root@other # Add to the local remote
kw remote remove origin --global
```

Could you also add a new commit for that?

Kamorst commented:
> That is a very nice update!
> 
> I added a few comments inline.
> 
> Btw, we need one option to indicate kw remote if it should update the global config or the local one where local is the default if nothing is used.
> 
> For example:
> 
> ```
> kw remote add origin root@something --global # update global config
> kw remote add origin root@other # Add to the local remote
> kw remote remove origin --global
> ```
> 
> Could you also add a new commit for that?

Added `--global` support to all remote sub-commands instead. I hope its okay.

I had forgotten to add this in a new commit, is that necessary?

rodrigosiqueira commented:
Merged! thanks
